src/coherence/nslookup.py

```diff
@@ -0,0 +1,293 @@
+            value = s[7]
+            result |= (value & 0x7F) << bits
+            bits += 7
+
+        if negative:
+            result = ~result
+
+        return result, bits
+
+    async def close(self) -> None:
+        if self.writer:
+            self.writer.close()
+            await self.writer.wait_closed()
+
+    async def discover_cluster_info(self) -> DiscoveredCluster:
+        cluster = DiscoveredCluster()
+        cluster.ns_port = self.port
+        cluster.host = self.host
+
+        cluster.cluster_name = await self.lookup(CLUSTER_NAME_LOOKUP)
+        cluster.management_urls = parse_results(await self.lookup(NS_PREFIX + MANAGEMENT_LOOKUP))
+        cluster.jmx_urls = parse_results(await self.lookup(NS_PREFIX + JMX_LOOKUP))
+        cluster.metrics_urls = parse_results(await self.lookup(NS_PREFIX + METRICS_LOOKUP))
+        cluster.grpc_proxy_endpoints = parse_results(await self.lookup(NS_PREFIX + GRPC_PROXY_LOOKUP))
+
+        return cluster
+
+    async def discover_name_service_ports(self) -> List[ClusterNSPort]:
+        local_cluster = await self.lookup(CLUSTER_NAME_LOOKUP)
+        other_clusters = await self.lookup(NS_PREFIX + CLUSTER_FOREIGN_LOOKUP)
+        other_clusters_list = parse_results(other_clusters)
+
+        cluster_names = [local_cluster] + other_clusters_list
+        list_clusters = [
+            ClusterNSPort(
+                cluster_name=name,
+                port=self.port if name == local_cluster else None,
+                host_name=self.host,
+                is_local=name == local_cluster,
+            )
+            for name in cluster_names
+        ]
+
+        for cluster_ns_port in list_clusters[1:]:
+            cluster_ns_port.port = int(
+                await self.lookup(f"{NS_PREFIX}{CLUSTER_FOREIGN_LOOKUP}/{cluster_ns_port.cluster_name}{NS_LOCAL_PORT}")
+            )
+
+        return list_clusters
+
+    @staticmethod
+    async def resolve_nslookup_address(name: str) -> str:
+
+        nslookup = None
+        cluster_info = None
+        try:
+            nslookup = await AsyncNSLookup.open(name)
+            cluster_info = await nslookup.discover_cluster_info()
+        except Exception as e:
+            print(f"Error: {e}")
+        finally:
+            if nslookup is not None:
+                await nslookup.close()
+
+        if cluster_info is not None:
+            # Combine the list into host:port pairs
+            pairs = [
+                f"{cluster_info.grpc_proxy_endpoints[i]}:{cluster_info.grpc_proxy_endpoints[i + 1]}"
+                for i in range(0, len(cluster_info.grpc_proxy_endpoints), 2)
+            ]
+            # Return a random pair
+            return random.choice(pairs)
+        else:
+            # return default target address
+            return "localhost:1408"
+
+
+def parse_results(results: str) -> List[str]:
+    results = results.strip("[]")
+    return results.split(", ") if results else []
+
+
+async def main() -> None:
```

Comment:
Does this need to be here now?  Perhaps add debug logging that includes the print statements within the relevant method.

Reply:
Its used by the implementation, so I am leaving it here